slick-postgres-driver/src/main/scala/org/hatdex/libs/dal/BaseSchemaMigrationImpl.scala

```diff
@@ -51,118 +51,76 @@ trait BaseSchemaMigrationImpl extends SchemaMigration {
-    Future(db.createSession().conn).map { dbConnection ⇒
-      changeLogFiles.foldLeft(Future.successful(())) { (execution, evolution) ⇒
-        execution.flatMap { _ ⇒
-          logger.error(s"Running evolution $evolution")
-          updateDb(evolution, dbConnection)
-        }
-      } onComplete {
-        _ ⇒ dbConnection.close()
+    logger.info(s"Running schema migrations: ${changeLogFiles.mkString(", ")}")
+    Future(db.createSession().conn).flatMap { dbConnection =>
+      val sequencedEvolutions: Future[Unit] = changeLogFiles.foldLeft(Future.successful(())) { (execution, evolution) =>
+        execution.flatMap(_ => updateDb(evolution, dbConnection))
       }
+      sequencedEvolutions.onComplete(_ => dbConnection.close())
+      sequencedEvolutions
     } recover {
-      case e ⇒ logger.error(s"Running database evolutions failed: ${e.getMessage}", e)
+      case e => logger.error("Running database evolutions failed", e)
     }
   }
 
-  def resetDatabase(): Future[Unit] = {
-    val eventuallyEvolved = Future {
-      val dbConnection = db.createSession().conn
-
-      val liquibase = blocking {
-        createLiquibase(dbConnection, "")
-      }
+  def resetDatabase(): Future[Unit] =
+    Future {
       blocking {
-        Try(liquibase.dropAll())
-          .recover {
-            case e =>
-              liquibase.forceReleaseLocks()
-              logger.error(s"Error dropping all database information")
-              throw e
-          }
-        liquibase.forceReleaseLocks()
+        val dbConnection = db.createSession().conn
+        val liquibase = createLiquibase(dbConnection, "")
+        try liquibase.dropAll()
+        catch {
+          case NonFatal(th) =>
+            logger.error("Error dropping all database information", th)
+            throw th
+        } finally liquibase.forceReleaseLocks()
       }
     }
 
-    eventuallyEvolved.failed.foreach { e =>
-      logger.error(s"Error updating database: ${e.getMessage}")
-    }
-
-    eventuallyEvolved
-  }
-
   def rollback(changeLogFiles: Seq[String]): Future[Unit] = {
     logger.info(s"Rolling back schema migrations: ${changeLogFiles.mkString(", ")}")
     changeLogFiles.foldLeft(Future.successful(())) { (execution, evolution) => execution.flatMap { _ => rollbackDb(evolution) } }
   }
 
-  private def updateDb(diffFilePath: String, dbConnection: Connection): Future[Unit] = {
-    val eventuallyEvolved = Future {
-
-      logger.info(s"Liquibase running evolutions $diffFilePath on db: [${dbConnection.getMetaData.getURL}]")
-      val liquibase = blocking {
-        createLiquibase(dbConnection, diffFilePath)
-      }
+  private def updateDb(diffFilePath: String, dbConnection: Connection): Future[Unit] =
+    Future {
       blocking {
+        logger.info(s"Liquibase running evolutions $diffFilePath on db: [${dbConnection.getMetaData.getURL}]")
+        val liquibase = createLiquibase(dbConnection, diffFilePath)
```

Comment:
everywhere else there is a session ```val dbConnection = db.createSession().conn``` created before createLiquibase is called but not here.

Reply:
because the same SQL connection is used multiple times by the one entry point:
`def run(changeLogFiles: Seq[String]): Future[Unit]`